File: app/src/main/java/com/iug/coursehub/feature/user/home_screen/course_details_screen/CourseDetailsFragment.java
package com.iug.coursehub.feature.user.home_screen.course_details_screen;

import static androidx.navigation.Navigation.findNavController;

import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.iug.coursehub.R;
import com.iug.coursehub.data.local.db.entity.Course;
import com.iug.coursehub.databinding.FragmentCourseDetailsBinding;

public class CourseDetailsFragment extends Fragment {

    private CourseDetailsViewModel viewModel;
    private Integer courseId;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            courseId = getArguments().getInt(ARG_COURSE_ID, 0);
        }
        viewModel = new ViewModelProvider(this).get(CourseDetailsViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        FragmentCourseDetailsBinding binding = FragmentCourseDetailsBinding.inflate(inflater, container, false);

        viewModel.getCourseDetails(courseId);

        viewModel.getUiAction().observe(getViewLifecycleOwner(), action -> {
            if (action instanceof CourseDetailsViewModel.UiAction.CollectCourseDetails) {
                CourseDetails courseDetails = ((CourseDetailsViewModel.UiAction.CollectCourseDetails) action).getCourseDetails();
                fillFields(binding, courseDetails);
            } else if (action instanceof CourseDetailsViewModel.UiAction.ShowToast) {
                Toast.makeText(requireContext(), ((CourseDetailsViewModel.UiAction.ShowToast) action).getMessage(), Toast.LENGTH_SHORT).show();
            }
        });

        binding.btnNavigateUp.setOnClickListener(v -> findNavController(requireView()).navigate(R.id.action_courseDetailsFragment_to_courseCategoriesFragment));
        binding.btnJoinCourse.setOnClickListener(v -> viewModel.joinCourse(courseId));

        return binding.getRoot();
    }

    private void fillFields(FragmentCourseDetailsBinding binding, CourseDetails courseDetails) {
        Course course = courseDetails.getCourse();
        byte[] imageByteArray = course.getImageByteArray();
        if (imageByteArray != null) {
            binding.courseImage.setImageBitmap(BitmapFactory.decodeByteArray(imageByteArray, 0, imageByteArray.length));
        }
        binding.courseName.setText(course.getName());
        binding.coursePrice.setText("Price: $" + course.getPrice());
        binding.courseStudents.setText("Students Enrolled: " + courseDetails.getNoOfEnrolledStudents());
        binding.courseHours.setText("Total Hours: " + course.getNoOfHours());
        binding.courseContents.setText("The course contains " + courseDetails.getNoOfLessons() + " lessons.\nJoin the course to see them.");
        binding.courseLecturer.setText("Lecturer: [" + course.getLecturerName() + "]");
    }

    public static final String ARG_COURSE_ID = "course_id";
}

File: app/src/main/java/com/iug/coursehub/feature/user/home_screen/course_details_screen/CourseDetailsFragment.kt
package com.iug.coursehub.feature.user.home_screen.course_details_screen

import android.graphics.BitmapFactory
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.iug.coursehub.R
import com.iug.coursehub.data.local.db.entity.Course
import com.iug.coursehub.databinding.FragmentCourseDetailsBinding

class CourseDetailsFragment : Fragment() {

    private val viewModel: CourseDetailsViewModel by viewModels()

    private var courseId: Int? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            courseId = it.getInt(ARG_COURSE_ID, 0)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = FragmentCourseDetailsBinding.inflate(inflater, container, false)

        binding.apply {
            viewModel.getCourseDetails(courseId!!)

            viewModel.uiAction.observe(viewLifecycleOwner) { action ->
                when(action) {
                    is CourseDetailsViewModel.UiAction.CollectCourseDetails -> {
                        val courseDetails = action.courseDetails
                        this.fillFields(courseDetails)
                    }

                    is CourseDetailsViewModel.UiAction.ShowToast -> {
                        Toast.makeText(requireContext(), action.message, Toast.LENGTH_SHORT).show()
                    }
                }
            }

            btnNavigateUp.setOnClickListener {
                findNavController().navigate(R.id.action_courseDetailsFragment_to_courseCategoriesFragment)
            }

            btnJoinCourse.setOnClickListener {
                viewModel.joinCourse(courseId!!)
            }
        }

        return binding.root
    }

    private fun FragmentCourseDetailsBinding.fillFields(courseDetails: CourseDetails) {
        val course = courseDetails.course
        val imgBitmap = BitmapFactory.decodeByteArray(course.imageByteArray, 0, course.imageByteArray.size)
        courseImage.setImageBitmap(imgBitmap)
        courseName.text = course.name
        coursePrice.text = "Price: $${course.price}"
        courseStudents.text = "Students Enrolled: ${courseDetails.noOfEnrolledStudents}"
        courseHours.text = "Total Hours: ${course.noOfHours}"
        courseContents.text = "The course contains ${courseDetails.noOfLessons} lessons.\nJoin the course to see them."
        courseLecturer.text = "Lecturer: [${course.lecturerName}]"
    }

    companion object {
        const val ARG_COURSE_ID = "course_id"
    }
}